### PR TITLE
Dependency update error processor test infrastructure

### PR DESCRIPTION
**Pull request by epananth**
This just sets up the test infrastructure. 

Does not test anything just the setup

### CODE REVIEW

File: src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            StateManager = new MockReliableStateManager();
+            RemoteFactory = new Mock<IRemoteFactory>(MockBehavior.Strict);
+            Env = new Mock<IHostingEnvironment>(MockBehavior.Strict);
+            services.AddSingleton(Env.Object);
+            services.AddSingleton<IReliableStateManager>(StateManager);
+            services.AddLogging();
+            services.AddDbContext<BuildAssetRegistryContext>(
+                options => { options.UseInMemoryDatabase("BuildAssetRegistry"); });
+            Provider = services.BuildServiceProvider();
+            services.Configure<GitHubTokenProviderOptions>(
```
Comment (epananth) [inline src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs:36]: Don't think I need this, want to verify this once.

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs [hunk 2]
```diff
@@ -58,15 +58,7 @@ IOptions<DependencyUpdateErrorProcessorOptions> options
...
-            if (_options.ConfigurationRefresherEndPointUri == null && _options.DynamicConfigs == null)
-            {
-                _logger.LogInformation("Dependency Update Error processor is disabled because no App Configuration was available.");
-                return;
-            }
-            await _options.ConfigurationRefresherEndPointUri.Refresh();
-            if (bool.TryParse(_options.DynamicConfigs["FeatureManagement:DependencyUpdateErrorProcessor"],
-                out var dependencyUpdateErrorProcessorFlag) &&
-                dependencyUpdateErrorProcessorFlag)
+            if(_options.IsEnabled)
```
Comment (epananth) [inline src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:61]: Removed this code here, cos Alex fixed the feature flag calling part. So I can directly call featureflag from config.

File: src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/CreateGithubIssueTests.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+                Timestamp = new DateTime(2020, 04, 01)
+            };
+            using (var context = new BuildAssetRegistryContext(Context.HostingEnvironment, options))
+            {
+                context.RepoBranchUpdateInMemory = new List<RepositoryBranchUpdateHistoryEntry>
+                    {repositoryBranchUpdate};
+                DependencyUpdateErrorProcessor errorProcessor =
+                    ActivatorUtilities.CreateInstance<DependencyUpdateErrorProcessor>(Scope.ServiceProvider, context);
+                await errorProcessor.ProcessDependencyUpdateErrorsAsync();
+            }
```
Comment (epananth) [inline src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/CreateGithubIssueTests.cs:34]: This test gets the mocked data I have set, so I waned to check if that worked. Cannot really verify anything here. I will add another PR just for test, next week.

File: src/Maestro/DependencyUpdateErrorProcessor/.config/settings.Production.json [hunk 4]
```diff
@@ -2,7 +2,5 @@
   "KeyVaultUri": "https://maestroprod.vault.azure.net/",
   "AppConfigurationUri": "https://maestroprod.azconfig.io/",
   "GithubUrl": "https://github.com/dotnet/core-eng",
-  "Owner": "dotnet",
```
Comment (riarenas) [inline src/Maestro/DependencyUpdateErrorProcessor/.config/settings.Production.json:5]: This doesn't seem related to the tests?

File: src/Maestro/DependencyUpdateErrorProcessor/.config/settings.json [hunk 5]
```diff
@@ -5,5 +5,6 @@
   "GitHub": {
     "GitHubAppId": "[vault(github-application-id)]",
     "PrivateKey": "[vault(github-application-private-key)]"
-  }
+  },
+  "EnableDependencyUpdateErrorProcessor": "[config(FeatureManagement:DependencyUpdateErrorProcessor)]"
```
Comment (riarenas) [inline src/Maestro/DependencyUpdateErrorProcessor/.config/settings.json:9]: Same with this change? Is it intentional?

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs (hunk 2 above)
Comment (riarenas) [reply src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:61]: I would make that a separate PR, Is it related to the test infrastructure?

File: src/Maestro/DependencyUpdateErrorProcessor/.config/settings.Production.json (hunk 4 above)
Comment (epananth) [reply src/Maestro/DependencyUpdateErrorProcessor/.config/settings.Production.json:5]: I previously had separate keys for owner and repo, I was not parsing the github url.  Now I did so we no longer need this. So I cleaned this up.

File: src/Maestro/DependencyUpdateErrorProcessor/.config/settings.json (hunk 5 above)
Comment (epananth) [reply src/Maestro/DependencyUpdateErrorProcessor/.config/settings.json:9]: Yes , Alex made some changes to the fetch the feature flag directly from the config. So I redid this part.  Should I add a small part saying I changed this in the issue?

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs (hunk 2 above)
Comment (epananth) [reply src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:61]: okay let me quickly create a new pr for that :)

File: src/Maestro/DependencyUpdateErrorProcessor/.config/settings.Production.json (hunk 4 above)
Comment (riarenas) [reply src/Maestro/DependencyUpdateErrorProcessor/.config/settings.Production.json:5]: I recommend adding these kinds of changes to the description of your PR.

Looking at the title and the description, you only mention that you're adding test infrastructure, so it's hard to know if a change that is not related to that is intentional or a merge problem

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs (hunk 2 above)
Comment (riarenas) [reply src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:61]: I think it's fine to keep it here, but it's probably a good idea to call out that you're adding these changes to the PR description.

I didn't know if you actually intended these to be part of the PR, or were left overs from local testing.

File: src/Maestro/DependencyUpdateErrorProcessor/.config/settings.Production.json (hunk 4 above)
Comment (epananth) [reply src/Maestro/DependencyUpdateErrorProcessor/.config/settings.Production.json:5]: I will create  a new PR for this. Sorry for the confusion

File: src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs [hunk 6]
```diff
@@ -77,6 +76,9 @@ public class BuildAssetRegistryContext : IdentityDbContext<ApplicationUser, Iden
         public DbSet<DependencyFlowEvent> DependencyFlowEvents { get; set; }
         public DbSet<GoalTime> GoalTime { get; set; }
         public DbSet<LongestBuildPath> LongestBuildPaths { get; set; }
+        // This is for testing purpose only. Instead of adding records to the db, this will let us add 
+        // record to the list.
+        public List<RepositoryBranchUpdateHistoryEntry> RepoBranchUpdateInMemory { get; set; }
```
Comment (riarenas) [inline src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs:81]: Is there a compelling reason to add this test property to the real class instead of creating a "TestBuildAssetRegistryContext"  class?

File: src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/CreateGithubIssueTests.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+                .Options;
+            var repositoryBranchUpdate = new RepositoryBranchUpdateHistoryEntry()
+            {
+                Repository = "https://github.com/maestro-auth-test/maestro-test2",
+                Branch = "38",
+                Method = "SynchronizePullRequestAsync",
+                Success = false,
+                Timestamp = new DateTime(2020, 04, 01)
+            };
+            using (var context = new BuildAssetRegistryContext(Context.HostingEnvironment, options))
```
Comment (riarenas) [inline src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/CreateGithubIssueTests.cs:27]: Why is it a new context and not getting the one that you built in the options?

Comment (riarenas) [reply src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/CreateGithubIssueTests.cs:27]: I don't think this is the right place for this test.

Once you create a test BuldAssetRegistry, this test should be part of the tests for that class.

File: src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs (hunk 1 above)
Comment (riarenas) [reply src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs:37]: We synced up, let me know if any questions are still open about this one.

File: src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs (hunk 6 above)
Comment (ChadNedzlek) [reply src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs:81]: I agree. The real code shouldn't be doing this.  This might mean something needs to be virtual, or some sort of inheritance, but the testing code shouldn't intrude into the real code.

File: src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/CreateGithubIssueTests.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+using Maestro.Data;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.DependencyInjection;
+using Xunit;
+
+namespace DependencyUpdateErrorProcessor.Tests
+{
+    public class CreateGithubIssueTest :DependencyUpdateErrorProcessorTests
+    {
+        [Fact]
```
Comment (ChadNedzlek) [inline src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/CreateGithubIssueTests.cs:13]: There don't appear to be any asserts in this fact... what is it supposed to be testing?

Comment (epananth) [reply src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/CreateGithubIssueTests.cs:13]: Chad, this is just for setting up the infrastructure. I;m separating the feature flag changes with the setting up the infrastructure. I will need add the tests in the next pr.

Comment (ChadNedzlek) [reply src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/CreateGithubIssueTests.cs:13]: I see... we probably shouldn't check the [Fact] in then, since it's not really a fact.

File: src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs (hunk 6 above)
Comment (alexperovich) [reply src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs:81]: The issue is the DbQuery configuration. the OnModelCreating method needs to configure the query to use for each DbQuery property. For the real implementation the DbQuery uses a raw SQL statement. The In-Memory provider can't handle that SQL, so we have to change how the query is set up. We might be able to just override the OnModelCreating in a new class and set up the query again with the testing implementation. I don't know if the model builder thingy can handle setting up the same query twice.

Comment (ChadNedzlek) [reply src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs:81]: The only thing that even uses this property is the "SaveAsync" method.  It should just call some virtual method that get's overridden by the mock.

Comment (ChadNedzlek) [reply src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs:81]: Sorry, being tricked by GitHub.  Whatever the giant method is that's using it... that code that is green here can just be moved into a virtual method that's overridden by the mock.

Comment (alexperovich) [reply src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs:81]: The thing that needs to be changed for the tests is the RepositoryBranchUpdateHistory property, I don't think EF would handle just overriding that property implementation to not be a normal property correctly.

Comment (ChadNedzlek) [reply src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs:81]: It's a new property we just added.  Add it to the Mock and not to the real class.

Comment (alexperovich) [reply src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs:81]: The new property is there only so the `builder.Query<RepositoryBranchUpdateHistoryEntry>()` call in OnModelCreating can use it. That is how the DbQuery properties are set up.

Comment (ChadNedzlek) [reply src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs:81]: If you are telling me that EF isn't testable, I guess that's fine, just one more thing to pile on my giant stack of reasons I dislike EF, but that seems unlikely to be true.  What would happen if this property were put in a sub class, and the code in green was made virtual?  Would it somehow just cause the huge pile of magic that is EF to burst into flame?

Comment (alexperovich) [reply src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs:81]: It might work if we put things in a sub class. I am not sure though.

Comment (epananth) [reply src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs:81]: I will try this and update this thread.

File: src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs (hunk 1 above)
Comment (epananth) [reply src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs:37]: okay I will

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs (hunk 2 above)
Comment (epananth) [reply src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs]: I moved that to a different pr. Resolving this

File: src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs (hunk 6 above)
Comment (epananth) [reply src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs:81]: I was able to extract this to a separate class. :)

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs [hunk 9]
```diff
@@ -55,10 +55,18 @@ IOptions<DependencyUpdateErrorProcessorOptions> options
             _logger = logger;
         }
 
-       [CronSchedule("0 0 0/1 1/1 * ? *", TimeZones.PST)]
-       public async Task ProcessDependencyUpdateErrorsAsync()
+        [CronSchedule("0 0 0/1 1/1 * ? *", TimeZones.PST)]
```
Comment (ChadNedzlek) [inline src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:58]: Why add a space to this line?

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs [hunk 10]
```diff
@@ -55,10 +55,18 @@ IOptions<DependencyUpdateErrorProcessorOptions> options
...
+        public async Task ProcessDependencyUpdateErrorsAsync()
         {
-            if (_options.IsEnabled)
```
Comment (ChadNedzlek) [inline src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:61]: We probably shouldn't undo your previous checkin. :-)

Comment (epananth) [reply src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:61]: Yes, my VS was acting weird and github was down. I have fixed that locally, will check in my changes tomorrow :)

File: src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs (hunk 6 above)
Comment (epananth) [reply src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs]: I added this as a separate class, so resolving this conversation.

File: src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/CreateGithubIssueTests.cs (hunk 8 above)
Comment (epananth) [reply src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/CreateGithubIssueTests.cs]: Moved this to setup code. I will add the test in another pr

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs (hunk 9 above)
Comment (epananth) [reply src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs]: redid this

File: arcade-services.sln [hunk 11]
```diff
@@ -140,554 +140,384 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.DotNet.Internal.L
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.DncEng.DeployServiceFabricCluster", "src\Microsoft.DncEng.DeployServiceFabricCluster\Microsoft.DncEng.DeployServiceFabricCluster.csproj", "{6307FF01-47DB-4EE6-86FF-9E78D8832D37}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Microsoft.DncEng.Configuration.Bootstrap", "src\Microsoft.DncEng.Configuration.Bootstrap\Microsoft.DncEng.Configuration.Bootstrap.csproj", "{2D72FBB5-9F29-40D5-9206-92A814DF47EC}"
```
Comment (riarenas) [inline arcade-services.sln:143]: Can we really remove this?

Comment (riarenas) [reply arcade-services.sln:143]: I'm guessing merge error?

Comment (alexperovich) [reply arcade-services.sln:143]: Yea, this looks like a merge issue.

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs [hunk 12]
```diff
@@ -42,17 +39,21 @@ internal sealed class DependencyUpdateErrorProcessor : IServiceImplementation
...
 
+        public Func<string , GitHubClient> AuthenticateGitHubClient { get; }
+
         public DependencyUpdateErrorProcessor(
             IReliableStateManager stateManager,
             ILogger<DependencyUpdateErrorProcessor> logger,
             BuildAssetRegistryContext context,
-            IOptions<DependencyUpdateErrorProcessorOptions> options
+            IOptions<DependencyUpdateErrorProcessorOptions> options,
+            Func<string, GitHubClient> authenticateGithubClient
```
Comment (riarenas) [inline src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:49]: Interesting way to pass the authentication haha. I guess it saves you the trouble of creating a mock class and injecting that instead of the real one during the tests?

File: arcade-services.sln (hunk 11 above)
Comment (ChadNedzlek) [reply arcade-services.sln:143]: Yeah, I had the same merge conflict. :-) sln files are the worst because of the "EndProject" lines, nothing can merge them automatically without screwing them up.

File: src/Maestro/DependencyUpdateErrorProcessor/Program.cs [hunk 13]
```diff
@@ -53,16 +52,36 @@ private static void Main()
...
-                                options.FyiHandle = config["FyiHandle"];
-                            });
+                                {
+                                    var config = provider.GetRequiredService<IConfiguration>();
+                                    options.IsEnabled = bool.Parse(config["EnableDependencyUpdateErrorProcessor"]);
+                                    options.GithubUrl = config["GithubUrl"];
+                                    options.FyiHandle = config["FyiHandle"];
+                                });
                         });
+                    host.ConfigureContainer(builder => builder.Register((c, p) =>
```
Comment (ChadNedzlek) [inline src/Maestro/DependencyUpdateErrorProcessor/Program.cs:64]: This won't with the validation I just checked in... this is going to fail your PR.  Please avoid the Autofac stuff unless you have no choice.

We definition shouldn't be starting a bunch of threads and waiting for them here.  This is bad.

File: src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+        public void Dispose()
+        {
+            Env.VerifyAll();
+            Scope.Dispose();
+            Provider.Dispose();
+        }
+
+        public void SetUp()
+        {
+            var repositoryBranchUpdate = new RepositoryBranchUpdateHistoryEntry()
```
Comment (ChadNedzlek) [inline src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs:73]: Why is there a bunch of code in the constructor AND a bunch of code in SetUp? It seems like only one of the two should be necessary.

File: src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+            };
+
+            var installationId = Context.GetInstallationId("TestValue");
+            string token = "TestToken";
+            Mock<GitHubAppTokenProvider> githubProvider = new Mock<GitHubAppTokenProvider>();
+            githubProvider.Setup(x => x.GetAppToken()).Returns(token);
+            Context.RepoBranchUpdateInMemory = new List<RepositoryBranchUpdateHistoryEntry>
+                {repositoryBranchUpdate};
+            DependencyUpdateErrorProcessor errorProcessor =
+                ActivatorUtilities.CreateInstance<DependencyUpdateErrorProcessor>(Scope.ServiceProvider,
```
Comment (ChadNedzlek) [inline src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs:89]: Any reason we don't use ServiceProvider.GetService<DependencyUpdateErrorProcessor> instead of the weird ActivatorUtilities thing?

File: src/Shared/Microsoft.DotNet.GitHub.Authentication/GitHubAppTokenProvider.cs [hunk 16]
```diff
@@ -17,7 +17,7 @@ public GitHubAppTokenProvider(IOptions<GitHubTokenProviderOptions> options = nul
             _options = options;
         }
 
-        public string GetAppToken()
+        public virtual string GetAppToken()
```
Comment (ChadNedzlek) [inline src/Shared/Microsoft.DotNet.GitHub.Authentication/GitHubAppTokenProvider.cs:20]: It might make more sent to extract an interface out of this than make a single method overridable.  This is the only method in the class... if you override it, none of the code in the class means anything anymore...

File: src/Maestro/DependencyUpdateErrorProcessor/Program.cs (hunk 13 above)
Comment (ChadNedzlek) [reply src/Maestro/DependencyUpdateErrorProcessor/Program.cs:64]: Actually, you're going to have a pretty massive merge conflict in this file (since I moved all the code in it), probably want to merge that in and see what's up.

Comment (alexperovich) [reply src/Maestro/DependencyUpdateErrorProcessor/Program.cs:64]: The only other way to do this is with a factory/provider interface. Like IGitHubClientFactory.

We use the autofac stuff so we can parameterize the resolution because we need a separate GitHubClient per repo. This can also be done with an interface and implementing class, but I didn't want to make a whole new class for this.

File: src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs (hunk 15 above)
Comment (alexperovich) [reply src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs:89]: The type isn't in the service collection, we could put it in there though.

File: src/Maestro/DependencyUpdateErrorProcessor/Program.cs (hunk 13 above)
Comment (ChadNedzlek) [reply src/Maestro/DependencyUpdateErrorProcessor/Program.cs:64]: Well, it's still not going to pass validation, so we don't have much choice now.  The tests won't be able to resolve the func.  We also _already_ have a class for this, so let's just use it.

File: src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs (hunk 15 above)
Comment (ChadNedzlek) [reply src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs:89]: Ooo, interesting.  I hadn't considered resolving something with the collection that wasn't in the collection.  I didn't even know that was possible.

File: src/Maestro/DependencyUpdateErrorProcessor/Program.cs (hunk 13 above)
Comment (ChadNedzlek) [reply src/Maestro/DependencyUpdateErrorProcessor/Program.cs:64]: I'd rather the code be in a nice, encapsulated, reusable chunk as well.  This isn't the only service that needs to interact with github, let's not make everyone copy paste a bunch of sync-over-async code into their startup methods.

Comment (ChadNedzlek) [reply src/Maestro/DependencyUpdateErrorProcessor/Program.cs:64]: Ooo, but it will. Because this thing isn't in the collection, it won't get validated...  that's... probably not good (and I'm going to fix that right quick, so don't depend on it).

Comment (ChadNedzlek) [reply src/Maestro/DependencyUpdateErrorProcessor/Program.cs:64]: Weird triple actually, there aren't any tests for this assembly, so I didn't add my test to them, so they aren't testing that their DI works at all.

We should add that test.

File: src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs (hunk 14 above)
Comment (epananth) [reply src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs:73]: Setup code will go to a Fact in my next PR. I didn't know where to put this code. So I added it as part of setup. Is that ok?

File: src/Maestro/DependencyUpdateErrorProcessor/Program.cs (hunk 13 above)
Comment (alexperovich) [reply src/Maestro/DependencyUpdateErrorProcessor/Program.cs:64]: > We also _already_ have a class for this

where is this class we already have? I don't remember it existing.

File: src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs (hunk 14 above)
Comment (ChadNedzlek) [reply src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs:73]: It's just odd the constructor is big, and this method is big.  Why not just have all the code in "SetUp"?

File: src/Maestro/DependencyUpdateErrorProcessor/Program.cs (hunk 13 above)
Comment (ChadNedzlek) [reply src/Maestro/DependencyUpdateErrorProcessor/Program.cs:64]: It's in a stupid place, src\DotNet.Status.Web\Controllers\GitHubClientFactory.cs.  It probably should get moved out to the Shared\*.GitHub.Authentication library.  I'm not sure what I didn't just put it there to start with, honestly.

Comment (epananth) [reply src/Maestro/DependencyUpdateErrorProcessor/Program.cs:64]: > Actually, you're going to have a pretty massive merge conflict in this file (since I moved all the code in it), probably want to merge that in and see what's up.

Hey @ChadNedzlek , I looked your commit https://github.com/dotnet/arcade-services/commit/c6a049cf6522bfb16a55a0c27b0e82ff046a7ef9, there is no change to my file in that.

Comment (ChadNedzlek) [reply src/Maestro/DependencyUpdateErrorProcessor/Program.cs:64]: Indeed, because there are no tests for it, so I didn't have to refactor it to make those tests...

I'm in the middle of removing all references to Autofac in our code, there will probably be a PR later today.  There is a bunch of missed validation because of it, and I'm trying to button that up.

Comment (epananth) [reply src/Maestro/DependencyUpdateErrorProcessor/Program.cs:64]: I m not blocked now, but I will check in my changes after I get your changes later today. So that there wont be any conflicts.

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs (hunk 12 above)
Comment (epananth) [reply src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:49]: Will no longer user this :D Will be just passing only the client :)

File: src/Shared/Microsoft.DotNet.GitHub.Authentication/GitHubAppTokenProvider.cs (hunk 16 above)
Comment (epananth) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/GitHubAppTokenProvider.cs:20]: As discussed creating another issue for this and will work on that issue as soon as this PR is merged. 

Issue -> https://github.com/dotnet/core-eng/issues/9582

File: src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs (hunk 14 above)
Comment (epananth) [reply src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs:73]: Since I will be moving the code inside my fact( in my next PR). I will just remove setup altogether, we don't actually need that code for now. I will put the setup in a different file in my next PR. That way we only have one constructor.

File: src/DotNet.Status.Web/Controllers/GitHubClientFactory.cs [hunk 17]
```diff
@@ -21,7 +21,7 @@ public GitHubClientFactory(IOptions<GitHubClientOptions> githubClientOptions, IG
             _tokenProvider = tokenProvider;
         }
 
-        public async Task<IGitHubClient> CreateGitHubClientAsync(string owner, string repo)
+        public virtual async Task<IGitHubClient> CreateGitHubClientAsync(string owner, string repo)
```
Comment (epananth) [inline src/DotNet.Status.Web/Controllers/GitHubClientFactory.cs:24]: Is it ok to make this method virtual?

File: src/Maestro/DependencyUpdateErrorProcessor/Program.cs (hunk 13 above)
Comment (epananth) [reply src/Maestro/DependencyUpdateErrorProcessor/Program.cs]: I checked in my changes with part of Chad's changes. I'm going to resolve this comment.

File: arcade-services.sln (hunk 11 above)
Comment (epananth) [reply arcade-services.sln]: This file change messes my entire project. It always screws up my commits. :(

File: src/DotNet.Status.Web/Controllers/GitHubClientFactory.cs (hunk 17 above)
Comment (ChadNedzlek) [reply src/DotNet.Status.Web/Controllers/GitHubClientFactory.cs:24]: Its the only method that does anything in this class.  If you make it virtual, the class becomes useless again.  And then you need to inject a bunch of options to let the constructor work, only to ignore them, which increases the complexity of the tests (having to configure a fake GitHubClientOptions and IGitHubTokenProvider) for no value gain.

It should get extracted into an interface if you want to mock something, and then mock the interface.

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs [hunk 18]
```diff
@@ -42,17 +40,21 @@ internal sealed class DependencyUpdateErrorProcessor : IServiceImplementation
         // github url -> https://github.com/maestro-auth-test/maestro-test2  repo ->maestro-test2 owner -> maestro-auth-test2 
         private static readonly Regex RepositoryUriPattern = new Regex(@"^/(?<owner>[^/]+)/(?<repo>[^/]+)/?$");
 
+        public GitHubClientFactory AuthenticateGitHubClient { get; }
```
Comment (ChadNedzlek) [inline src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:43]: public? All the other fields are private. Other things shouldn't be reaching into this object to get it's GitHubClientFactory.

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.csproj [hunk 19]
```diff
@@ -28,6 +28,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <ProjectReference Include="..\..\DotNet.Status.Web\DotNet.Status.Web.csproj" />
```
Comment (ChadNedzlek) [inline src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.csproj:31]: Don't reference the entire web project.  This needs to get moved to a different assembly. Probably the current GitHub.Authentication one. That's where it makes the most sense.

File: src/DotNet.Status.Web/Controllers/GitHubClientFactory.cs (hunk 17 above)
Comment (epananth) [reply src/DotNet.Status.Web/Controllers/GitHubClientFactory.cs:24]: @ChadNedzlek I am going to add this one also to the same issue I created yesterday. So I can work on it after I merge my PR. Is that fine?

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs (hunk 18 above)
Comment (epananth) [reply src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:43]: Oops missed this one. Changing this.

File: src/DotNet.Status.Web/Controllers/GitHubClientFactory.cs (hunk 17 above)
Comment (ChadNedzlek) [reply src/DotNet.Status.Web/Controllers/GitHubClientFactory.cs:24]: You're just making more work for yourself and introducing potential bugs.   It's a pretty small change, and I'd rather we don't check code we don't like into master, since it just increases the technical debt.  This change is so small that without it, this change introduces more debt than it fixes...

Comment (epananth) [reply src/DotNet.Status.Web/Controllers/GitHubClientFactory.cs:24]: Sure I will add this to the current PR.

File: src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+        protected readonly Mock<GitHubClientFactory> GithubClientFactory;
+
+        public DependencyUpdateErrorProcessorTests()
+        {
+            var services = new ServiceCollection();
+            StateManager = new MockReliableStateManager();
+            Env = new Mock<IHostingEnvironment>(MockBehavior.Strict);
+            GithubClient = new Mock<IGitHubClient>();
+            GithubClientFactory = new Mock<GitHubClientFactory>(null, null);
+            GithubClientFactory.Setup(x => x.CreateGitHubClientAsync("Owner", "Repo"))
```
Comment (ChadNedzlek) [inline src/Maestro/tests/DependencyUpdateErrorProcessor.Tests/DepenedencyUpdateErrorProcessorTests.cs:34]: Do you really only want it to return on these 2 values? If so, they should probably be constants, so that other tests can be sure the pass the right answer.

If you just want it to always return that, then use It.IsAny<string>()

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.csproj (hunk 19 above)
Comment (epananth) [reply src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.csproj]: Removed this reference and moved GithubClientFactory to Microsoft.DotNet.GitHub.Authentication proj

File: src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs [hunk 21]
```diff
@@ -42,17 +41,21 @@ internal sealed class DependencyUpdateErrorProcessor : IServiceImplementation
         // github url -> https://github.com/maestro-auth-test/maestro-test2  repo ->maestro-test2 owner -> maestro-auth-test2 
         private static readonly Regex RepositoryUriPattern = new Regex(@"^/(?<owner>[^/]+)/(?<repo>[^/]+)/?$");
 
+        private IGitHubClientFactory AuthenticateGitHubClient { get; }
```
Comment (ChadNedzlek) [inline src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs:44]: private readonly field, please.
